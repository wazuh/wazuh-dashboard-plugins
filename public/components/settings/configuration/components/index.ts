/*
 * Wazuh app - React component building the configuration component.
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 */
export { Header } from './header';
export { Categories } from './categories';
export { BottomBar } from './bottom-bar';