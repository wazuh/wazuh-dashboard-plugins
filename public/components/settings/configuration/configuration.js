/*
 * Wazuh app - React component building the configuration component.
 *
 * Copyright (C) 2015-2020 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 */

import React, { Component } from 'react';
import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiButtonIcon,
  EuiToolTip,
  EuiPanel,
  EuiPage,
  EuiTitle,
  EuiText
} from '@elastic/eui';
import { WzConfigurationTable } from './configuration-table';

export class WzConfigurationSettings extends Component {
  constructor(props) {
    super(props);

    this.state = {
      apiEntries: [],
      refreshingEntries: false
    };
  }

  componentDidMount() {}

  render() {
    return (
      <EuiPage>
        <EuiPanel paddingSize="l">
          <EuiFlexGroup>
            <EuiFlexItem>
              <EuiFlexGroup>
                <EuiFlexItem>
                  <EuiTitle>
                    <h2>
                      App current settings&nbsp;
                      <EuiToolTip
                        position="right"
                        content="More about configuration file"
                      >
                        <EuiButtonIcon
                          iconType="questionInCircle"
                          iconSize="l"
                          aria-label="Help"
                          target="_blank"
                          href="https://documentation.wazuh.com/current/user-manual/kibana-app/reference/config-file.html"
                        ></EuiButtonIcon>
                      </EuiToolTip>
                    </h2>
                  </EuiTitle>
                </EuiFlexItem>
              </EuiFlexGroup>
            </EuiFlexItem>
          </EuiFlexGroup>
          <EuiFlexGroup>
            <EuiFlexItem>
              <EuiText color="subdued" style={{ paddingBottom: '15px' }}>
                Configuration file located at
                /usr/share/kibana/optimize/wazuh/config/wazuh.yml
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
          <WzConfigurationTable {...this.props} />
        </EuiPanel>
      </EuiPage>
    );
  }
}
