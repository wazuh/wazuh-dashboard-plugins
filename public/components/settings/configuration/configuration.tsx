/*
 * Wazuh app - React component building the configuration component.
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 */

import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Header, BottomBar } from './components';
import { useKbnLoadingIndicator} from '../../common/hooks';
import { useForm } from '../../common/form/hooks';
import {
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
  EuiPage,
  EuiPageBody,
  EuiPageHeader,
  EuiSpacer,
  Query,
} from '@elastic/eui';
import store from '../../../redux/store'
import { updateSelectedSettingsSection } from '../../../redux/actions/appStateActions';
import { withUserAuthorizationPrompt, withErrorBoundary, withReduxProvider } from '../../common/hocs';
import { EpluginSettingType, PLUGIN_SETTINGS, PLUGIN_SETTINGS_CATEGORIES, UI_LOGGER_LEVELS, WAZUH_ROLE_ADMINISTRATOR_NAME } from '../../../../common/constants';
import { compose } from 'redux';
import { getPluginSettingDescription, getSettingsDefaultList, groupSettingsByCategory, getCategorySettingByTitle } from '../../../../common/services/settings';
import { Category } from './components/categories/components';
import { WzRequest } from '../../../react-services';
import { UIErrorLog, UIErrorSeverity, UILogLevel, UI_ERROR_SEVERITIES } from '../../../react-services/error-orchestrator/types';
import { getErrorOrchestrator } from '../../../react-services/common-services';
import { updateAppConfig } from '../../../redux/actions/appConfigActions';
import path from 'path';
import { toastRequiresReloadingBrowserTab, toastRequiresRestartingPluginPlatform, toastRequiresRunningHealthcheck, toastSuccessUpdateConfiguration } from './components/categories/components/show-toasts';

export type ISetting = {
  key: string
  value: boolean | string | number | object
  description: string
  category: string
  name: string
  readonly?: boolean
  form: { type: string, params: {} }
};

const pluginSettingConfigurableUI = getSettingsDefaultList()
  .filter(categorySetting => categorySetting.isConfigurableFromUI)
  .map(setting => ({ ...setting, category: PLUGIN_SETTINGS_CATEGORIES[setting.category].title}));

const settingsCategoriesSearchBarFilters = [...new Set(pluginSettingConfigurableUI.map(({category}) => category))].sort().map(category => ({value: category}))

const trasnsfromPluginSettingsToFormFields = configuration => Object.fromEntries(
  getSettingsDefaultList()
    .filter(pluginSetting => pluginSetting.isConfigurableFromUI)
    .map(({
      key,
      type,
      validate,
      defaultValue: initialValue,
      uiFormTransformChangedInputValue,
      uiFormTransformConfigurationValueToInputValue,
      uiFormTransformInputValueToConfigurationValue,
      ...rest
    }) => ([
      key,
      {
        type,
        validate: validate?.bind?.(rest),
        transformChangedInputValue: uiFormTransformChangedInputValue?.bind?.(rest),
        transformChangedOutputValue: uiFormTransformInputValueToConfigurationValue?.bind?.(rest),
        initialValue: uiFormTransformConfigurationValueToInputValue ? uiFormTransformConfigurationValueToInputValue.bind(rest)(configuration?.[key] ?? initialValue) : (configuration?.[key] ?? initialValue)
      }
    ]))
);

const WzConfigurationSettingsProvider = (props) => {
  const [loading, setLoading ] = useKbnLoadingIndicator();
  const [query, setQuery] = useState('');
  const currentConfiguration = useSelector(state => state.appConfig.data);

  const { fields, changed, errors, doChanges, undoChanges } = useForm(trasnsfromPluginSettingsToFormFields(currentConfiguration));
  const dispatch = useDispatch();

  useEffect(() => {
    store.dispatch(updateSelectedSettingsSection('configuration'));
  }, []);

  const onChangeSearchQuery = (query) => {
    setQuery(query);
  };

  const visibleSettings = Object.entries(fields)
    .map(([fieldKey, fieldForm]) => ({
      ...fieldForm,
      key: fieldKey,
      category: PLUGIN_SETTINGS_CATEGORIES[PLUGIN_SETTINGS[fieldKey].category].title,
      type: PLUGIN_SETTINGS[fieldKey].type,
      options: PLUGIN_SETTINGS[fieldKey]?.options,
      title: PLUGIN_SETTINGS[fieldKey]?.title,
      description: getPluginSettingDescription(PLUGIN_SETTINGS[fieldKey]),
    }));

  // https://github.com/elastic/eui/blob/aa4cfd7b7c34c2d724405a3ecffde7fe6cf3b50f/src/components/search_bar/query/query.ts#L138-L163
  const search = Query.execute(query.query || query, visibleSettings, ['description', 'key', 'title']);

  const visibleCategories = groupSettingsByCategory(search || visibleSettings)
  // Sort categories to render them in their enum definition order
    .sort((a, b) =>
      (getCategorySettingByTitle(a.category)?.renderOrder || 0) -
      (getCategorySettingByTitle(b.category)?.renderOrder || 0)
    );

  const onSave = async () => {
    setLoading(true);
    try {
      const settingsToUpdate = Object.entries(changed).reduce((accum, [pluginSettingKey, currentValue]) => {
        if(PLUGIN_SETTINGS[pluginSettingKey].isConfigurableFromFile && PLUGIN_SETTINGS[pluginSettingKey].type === EpluginSettingType.filepicker){
          accum.fileUpload = {
            ...accum.fileUpload,
            [pluginSettingKey]: {
              file: currentValue,
              extension: path.extname(currentValue.name)
            }
          }
        }else if(PLUGIN_SETTINGS[pluginSettingKey].isConfigurableFromFile){
          accum.saveOnConfigurationFile = {
            ...accum.saveOnConfigurationFile,
            [pluginSettingKey]: currentValue
          }
        };
        return accum;
      }, {saveOnConfigurationFile: {}, fileUpload: {}});

      const requests = [];

      // Update the settings that doesn't upload a file
      if(Object.keys(settingsToUpdate.saveOnConfigurationFile).length){
        requests.push(WzRequest.genericReq(
          'PUT', '/utils/configuration',
          settingsToUpdate.saveOnConfigurationFile
        ));
      };

      // Update the settings that uploads a file
      if(Object.keys(settingsToUpdate.fileUpload).length){
        requests.push(...Object.entries(settingsToUpdate.fileUpload)
          .map(([pluginSettingKey, {file}]) => {
              // Create the form data
              const formData = new FormData();
              formData.append('file', file);
              return WzRequest.genericReq(
                'PUT', `/utils/configuration/files/${pluginSettingKey}`,
                formData,
                {overwriteHeaders: {'content-type': 'multipart/form-data'}}
              )
            }));
      };

      const responses = await Promise.all(requests);

      // Show the toasts if necessary
      responses.some(({data: { data: {requiresRunningHealthCheck}}}) => requiresRunningHealthCheck) && toastRequiresRunningHealthcheck();
      responses.some(({data: { data: {requiresReloadingBrowserTab}}}) => requiresReloadingBrowserTab) && toastRequiresReloadingBrowserTab();
      responses.some(({data: { data: {requiresRestartingPluginPlatform}}}) => requiresRestartingPluginPlatform) && toastRequiresRestartingPluginPlatform();

      // Update the app configuration frontend-cached setting in memory with the new values
      const updatedConfiguration = {
        ...responses.reduce((accum, {data: {data}}) => {
          return {
            ...accum,
            ...(data.updatedConfiguration ? {...data.updatedConfiguration} : {}),
          }
        },{})
      };
      dispatch(updateAppConfig(updatedConfiguration));

      // Remove the selected files on the file picker inputs
      if(Object.keys(settingsToUpdate.fileUpload).length){
        Object.keys(settingsToUpdate.fileUpload).forEach(settingKey => {
          try{
            fields[settingKey].inputRef.removeFiles(
              // This method uses some methods of a DOM event.
              // Because we want to remove the files when the configuration is saved,
              // there is no event, so we create a object that contains the
              // methods used to remove the files. Of this way, we can skip the errors
              // due to missing methods.
              // This workaround is based in @elastic/eui v29.3.2
              // https://github.com/elastic/eui/blob/v29.3.2/src/components/form/file_picker/file_picker.tsx#L107-L108
              {stopPropagation: () => {}, preventDefault: () => {}}
            );
          }catch(error){ };
        });
      };

      // Show the success toast
      toastSuccessUpdateConfiguration();

      // Reset the form changed configuration
      doChanges();
    } catch (error) {
      const options: UIErrorLog = {
        context: `${WzConfigurationSettingsProvider.name}.onSave`,
        level: UI_LOGGER_LEVELS.ERROR as UILogLevel,
        severity: UI_ERROR_SEVERITIES.BUSINESS as UIErrorSeverity,
        store: true,
        error: {
          error: error,
          message: error.message || error,
          title: `Error saving the configuration: ${error.message || error}`,
        },
      };

      getErrorOrchestrator().handleError(options);
    } finally {
      setLoading(false);
    };
  };

  const onCancel = () => {
    const updatedSettingsUseFilePicker = Object.entries(changed).reduce((accum, [pluginSettingKey]) => {
      if(PLUGIN_SETTINGS[pluginSettingKey].isConfigurableFromFile && PLUGIN_SETTINGS[pluginSettingKey].type === EpluginSettingType.filepicker){
        accum.push(pluginSettingKey);
      };
      return accum;
    }, []);

    updatedSettingsUseFilePicker.forEach(settingKey => {
      try{
        fields[settingKey].inputRef.removeFiles(
          // This method uses some methods of a DOM event.
          // Because we want to remove the files when the configuration is saved,
          // there is no event, so we create a object that contains the
          // methods used to remove the files. Of this way, we can skip the errors
          // due to missing methods.
          // This workaround is based in @elastic/eui v29.3.2
          // https://github.com/elastic/eui/blob/v29.3.2/src/components/form/file_picker/file_picker.tsx#L107-L108
          {stopPropagation: () => {}, preventDefault: () => {}}
        );
      }catch(error){ };
    });
    undoChanges();
  };

  return (
    <EuiPage >
      <EuiPageBody className='mgtPage__body' restrictWidth>
        <EuiPageHeader>
          <Header
            query={query}
            setQuery={onChangeSearchQuery}
            searchBarFilters={[{
              type: 'field_value_selection',
              field: 'category',
              name: 'Categories',
              multiSelect: 'or',
              options: settingsCategoriesSearchBarFilters,
            }]}/>
        </EuiPageHeader>
        <EuiFlexGroup direction='column'>
          {visibleCategories && visibleCategories.map(({ category, settings }) => {
            const { description, documentationLink } = getCategorySettingByTitle(category);
            return (
              <Category
                key={`configuration_category_${category}`}
                title={category}
                description={description}
                documentationLink={documentationLink}
                items={settings}
                currentConfiguration={currentConfiguration}
              />
            )
          }
          )}
        </EuiFlexGroup>
        <EuiSpacer size="xxl" />
        {Object.keys(changed).length > 0 && (
          <BottomBar
            errorsCount={Object.keys(errors).length}
            unsavedCount={Object.keys(changed).length}
            onCancel={onCancel}
            onSave={onSave}
          />
        )}
      </EuiPageBody>
    </EuiPage>
  );
}
export const WzConfigurationSettings = compose (
  withErrorBoundary,
  withReduxProvider,
  withUserAuthorizationPrompt(null, [WAZUH_ROLE_ADMINISTRATOR_NAME])
)(WzConfigurationSettingsProvider);
