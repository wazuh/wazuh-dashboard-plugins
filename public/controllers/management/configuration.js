/*
 * Wazuh app - Management configuration controller
 * Copyright (C) 2015-2019 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 */
import { ConfigurationHandler } from '../../utils/config-handler';
import { DynamicHeight } from '../../utils/dynamic-height';

export class ConfigurationController {
  /**
   * Constructor
   * @param {*} $scope
   * @param {*} $location
   * @param {*} errorHandler
   * @param {*} apiReq
   * @param {*} appState
   */
  constructor($scope, $location, errorHandler, apiReq, appState) {
    this.$scope = $scope;
    this.errorHandler = errorHandler;
    this.apiReq = apiReq;
    this.appState = appState;
    this.$location = $location;
    this.load = false;
    this.isArray = Array.isArray;
    this.configurationHandler = new ConfigurationHandler(apiReq, errorHandler);
    this.currentConfig = null;
    this.configurationTab = '';
    this.configurationSubTab = '';
    this.integrations = {};
    this.$scope.integrations = {};
    this.$scope.selectedItem = 0;
    this.showHelp = false;
  }

  /**
   * When controller loads
   */
  $onInit() {
    this.$scope.getXML = () => this.configurationHandler.getXML(this.$scope);

    this.$scope.getJSON = () => this.configurationHandler.getJSON(this.$scope);

    this.$scope.switchConfigTab = (
      configurationTab,
      sections,
      navigate = true
    ) => this.switchConfigTab(configurationTab, sections, navigate);

    this.$scope.switchConfigurationSubTab = configurationSubTab =>
      this.switchConfigurationSubTab(configurationSubTab);

    this.$scope.switchWodle = (wodleName, navigate = true) =>
      this.switchWodle(wodleName, navigate);

    this.$scope.isString = item => typeof item === 'string';

    this.$scope.hasSize = obj =>
      obj && typeof obj === 'object' && Object.keys(obj).length;

    this.$scope.updateSelectedItem = i => (this.$scope.selectedItem = i);

    this.getIntegration = list =>
      this.configurationHandler.getIntegration(list, this.$scope);

    this.$scope.$on('$routeChangeStart', () =>
      this.appState.removeSessionStorageItem('configSubTab')
    );

    this.$scope.$on('configurationIsReloaded', (event, params) => {
      this.reloadConfig(params);
    });

    this.$scope.$on('configNodeChanged', () => {
      if (!this.configurationTab || !this.configurationTab.length) {
        this.configurationTab = 'welcome';
      }
      this.switchConfigurationTab(this.configurationTab, false, true);
    });

    this.$scope.configurationTabsProps = {};
    this.$scope.buildProps = tabs => {
      this.$scope.configurationTabsProps = {
        clickAction: tab => {
          this.$scope.switchConfigurationSubTab(tab);
        },
        selectedTab:
          this.$scope.configurationSubTab || (tabs && tabs.length)
            ? tabs[0].id
            : '',
        tabs
      };
    };

    $(window).on('resize', () => {
      DynamicHeight.dynamicHeight('d-height', 50);
    });
  }

  switchConfigurationSubTab(configurationSubTab) {
    this.$scope.selectedItem = 0;
    this.configurationHandler.switchConfigurationSubTab(
      configurationSubTab,
      this.$scope
    );

    if (configurationSubTab === 'pm-sca') {
      this.$scope.currentConfig.sca = this.configurationHandler.parseWodle(
        this.$scope.currentConfig,
        'sca'
      );
    }
    DynamicHeight.dynamicHeight('d-height', 50);
  }

  /**
   * Navigate to woodle
   */
  switchWodle(wodleName, navigate = true) {
    this.appState.setNavigation({ status: true });
    this.navigate = navigate;
    this.configWodle = wodleName;
    if (!this.$location.search().configWodle) {
      this.$scope.$emit('setCurrentConfiguration', {
        currentConfiguration: wodleName
      });
      this.$location.search('configWodle', this.configWodle);
    }
    this.configurationHandler.switchWodle(
      wodleName,
      this.$scope,
      false,
      (this.$scope.mctrl || {}).selectedNode
    );
    DynamicHeight.dynamicHeight('d-height', 50);
  }

  /**
   * Navigate to configuration
   */
  switchConfigurationTab(configurationTab, navigate) {
    this.$scope.mctrl.editionTab = '';
    this.navigate = navigate;
    this.configurationHandler.switchConfigurationTab(
      configurationTab,
      this.$scope
    );

    if (!this.navigate) {
      let configSubTab = this.$location.search().configSubTab;
      if (configSubTab) {
        try {
          const config = this.appState.getSessionStorageItem('configSubTab');
          const configSubTabObj = JSON.parse(config);
          if (!configSubTabObj) return;
          this.$scope.$emit('setCurrentConfiguration', {
            currentConfiguration: configSubTabObj.configurationTab
          });
          this.switchConfigTab(
            configSubTabObj.configurationTab,
            configSubTabObj.sections,
            false
          );
        } catch (error) {
          this.errorHandler.handle(error.message || error);
        }
      } else {
        let configWodle = this.$location.search().configWodle;
        if (configWodle) {
          this.$scope.$emit('setCurrentConfiguration', {
            currentConfiguration: configWodle
          });
          this.switchWodle(configWodle, false);
        }
      }
    } else {
      this.$location.search('configSubTab', null);
      this.appState.removeSessionStorageItem('configSubTab');
      this.$location.search('configWodle', null);
    }
  }

  reloadConfig(params) {
    if ((params || {}).globalConfigTab) {
      this.configurationTab = '';
      this.$scope.mctrl.editionTab = params.globalConfigTab;
      if ((params || {}).reloadConfigSubTab) {
        this.$location.search('configSubTab', null);
        this.appState.removeSessionStorageItem('configSubTab');
      }
      this.$scope.$emit('removeCurrentConfiguration', {});
    } else {
      this.$scope.mctrl.editionTab = '';
      this.switchConfigurationTab('welcome', true);
    }
    this.$scope.$applyAsync();
  }

  switchConfigTab(configurationTab, sections, navigate = true) {
    this.appState.setNavigation({ status: true });
    this.navigate = navigate;
    try {
      this.configSubTab = JSON.stringify({ configurationTab, sections });
      if (!this.$location.search().configSubTab) {
        this.appState.setSessionStorageItem('configSubTab', this.configSubTab);
        this.$scope.$emit('setCurrentConfiguration', {
          currentConfiguration: configurationTab
        });
        this.$location.search('configSubTab', true);
      }
      DynamicHeight.dynamicHeight('d-height', 50);
    } catch (error) {
      this.errorHandler.handle(error, 'Set configuration path');
    }

    this.configurationHandler.switchConfigTab(
      configurationTab,
      sections,
      this.$scope,
      false,
      (this.$scope.mctrl || {}).selectedNode
    );
  }
}
