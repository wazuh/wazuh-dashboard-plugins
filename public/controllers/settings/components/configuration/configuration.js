/*
 * Wazuh app - React component building the configuration component.
 *
 * Copyright (C) 2015-2019 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 */

import React, { Component } from 'react';
import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiButtonIcon,
  EuiToolTip,
  EuiPanel,
  EuiPage,
  EuiTitle,
  EuiText,
} from '@elastic/eui';
import { WzConfigurationTable } from './configuration-table';

export class WzConfigurationSettings extends Component {
  constructor(props) {
    super(props);

    this.state = {
      apiEntries: [],
      refreshingEntries: false,
    };
  }

  componentDidMount() {
    // this.setState({
    //   apiEntries: this.props.apiEntries
    // });
  }

  /**
   * Refresh the API entries
   */
  // async refresh() {
  //   try {
  //     let status = 'complete';
  //     this.setState({ error: false });
  //     const hosts = await this.props.getHosts();
  //     this.setState({
  //       refreshingEntries: true,
  //       apiEntries: hosts
  //     });
  //     const entries = this.state.apiEntries;
  //     let numErr = 0;
  //     for (let idx in entries) {
  //       const entry = entries[idx];
  //       try {
  //         const data = await this.props.testApi(entry);
  //         const clusterInfo = data.data || {};
  //         const id = entries[idx].id;
  //         entries[idx].status = 'online';
  //         entries[idx].cluster_info = clusterInfo;
  //         //Updates the cluster info in the registry
  //         await this.props.updateClusterInfoInRegistry(id, clusterInfo);
  //       } catch (error) {
  //         numErr = numErr + 1;
  //         const code = ((error || {}).data || {}).code;
  //         const downReason =
  //           ((error || {}).data || {}).message || 'Wazuh is not reachable';
  //         const status = code === 3099 ? 'down' : 'unknown';
  //         entries[idx].status = { status, downReason };
  //       }
  //     }
  //     if (numErr) {
  //       if (numErr >= entries.length) this.props.showApiIsDown();
  //     }
  //     this.setState({
  //       apiEntries: entries,
  //       status: status,
  //       refreshingEntries: false
  //     });
  //   } catch (error) {
  //     if (
  //       error &&
  //       error.data &&
  //       error.data.message &&
  //       error.data.code === 2001
  //     ) {
  //       this.props.showAddApiWithInitialError(error);
  //     }
  //   }
  // }

  /**
   * Checks the API connectivity
   * @param {Object} api
   */
  // async checkApi(api) {
  //   try {
  //     const entries = this.state.apiEntries;
  //     const idx = entries.map(e => e.id).indexOf(api.id);
  //     try {
  //       await this.props.checkManager(api);
  //       entries[idx].status = 'online';
  //     } catch (error) {
  //       const code = ((error || {}).data || {}).code;
  //       const downReason =
  //         ((error || {}).data || {}).message || 'Wazuh is not reachable';
  //       const status = code === 3099 ? 'down' : 'unknown';
  //       entries[idx].status = { status, downReason };
  //     }
  //     this.setState({
  //       apiEntries: entries
  //     });
  //   } catch (error) {
  //     console.error('Error checking manager connection ', error);
  //   }
  // }

  render() {
    return (
      <EuiPage>
        <EuiPanel paddingSize="l">
          <EuiFlexGroup>
            <EuiFlexItem>
              <EuiFlexGroup>
                <EuiFlexItem>
                  <EuiTitle>
                    <h2>
                      App current settings
                      <EuiToolTip position="right" content="More about configuration file">
                        <EuiButtonIcon
                          iconType="questionInCircle"
                          iconSize="l"
                          aria-label="Help"
                          target="_blank"
                          href="https://documentation.wazuh.com/current/user-manual/kibana-app/reference/config-file.html"
                        ></EuiButtonIcon>
                      </EuiToolTip>
                    </h2>
                  </EuiTitle>
                </EuiFlexItem>
              </EuiFlexGroup>
            </EuiFlexItem>
          </EuiFlexGroup>
          <EuiFlexGroup>
            <EuiFlexItem>
              <EuiText color="subdued" style={{ paddingBottom: '15px' }}>
                Configuration file located at /usr/share/kibana/plugins/wazuh/wazuh.yml
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
          <WzConfigurationTable {...this.props} />
        </EuiPanel>
      </EuiPage>
    );
  }
}
