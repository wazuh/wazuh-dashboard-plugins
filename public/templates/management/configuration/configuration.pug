include ./configuration.head
include ./welcome.html
include ./global-configuration/global-configuration.pug
include ./cluster/cluster.html
include ./config-edition/config-edition.html
include ./registration-service/registration-service.html
include ./alerts/alerts.pug
include ./integrations/integrations.html
include ./policy-monitoring/policy-monitoring.pug
include ./open-scap/open-scap.pug
include ./cis-cat/cis-cat.pug
include ./vulnerabilities/vulnerabilities.pug
include ./osquery/osquery.html
include ./inventory/inventory.html
include ./active-response/active-response.pug
include ./wazuh-commands/wazuh-commands.html
include ./log-collection/log-collection.pug
include ./integrity-monitoring/integrity-monitoring.pug
include ./agentless/agentless.html
include ./amazon-s3/amazon-s3.pug
include ./docker-listener/docker-listener.html
include ../../footer.foot