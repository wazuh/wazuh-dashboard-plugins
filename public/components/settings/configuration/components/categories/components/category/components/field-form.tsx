/*
 * Wazuh app - React component building the configuration component.
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 */
import React, { useState, useEffect } from 'react';
import { validate } from 'node-cron';
import {
  EuiFieldText,
  EuiFieldNumber,
  EuiSwitch,
  EuiSelect,
  EuiCodeEditor,
  EuiTextColor
} from '@elastic/eui';
import { ISetting } from '../../../../../configuration';
import 'brace/mode/javascript';
import 'brace/snippets/javascript';
import 'brace/ext/language_tools';
import "brace/ext/searchbox";
import {
  UI_ERROR_SEVERITIES,
  UIErrorLog, UIErrorSeverity,
  UILogLevel,
} from '../../../../../../../../react-services/error-orchestrator/types';
import { UI_LOGGER_LEVELS } from '../../../../../../../../../common/constants';
import { getErrorOrchestrator } from '../../../../../../../../react-services/common-services';
import _ from 'lodash';

interface IFieldForm {
  item: ISetting
  updatedConfig: { [field: string]: string | number | boolean | [] }
  setUpdatedConfig({ }): void
}
export const FieldForm: React.FunctionComponent<IFieldForm> = (props) => {
  const { item } = props;
  switch (item.form.type) {
    case 'text':
      return <TextForm {...props} />
    case 'number':
      return <NumberForm {...props} />
    case 'boolean':
      return <BooleanForm {...props} />
    case 'list':
      return <ListForm {...props} />
    case 'array':
      return <ArrayForm {...props} />
    case 'interval':
      return <IntervalForm {...props} />
    default:
      return null;
  }
};

//#region forms
const TextForm: React.FunctionComponent<IFieldForm> = (props) => (
  <EuiFieldText
    fullWidth
    value={getValue(props)}
    onChange={e => onChange(e.target.value, props)} />);

const NumberForm: React.FunctionComponent<IFieldForm> = (props) => (
  <EuiFieldNumber
    fullWidth
    value={getValue(props)}
    onChange={e => onChange(e.target.value, props)} />)
const BooleanForm: React.FunctionComponent<IFieldForm> = (props) => (
  <EuiSwitch
    label={`${getValue(props)}`}
    checked={getValue(props)}
    onChange={(e) => onChange(e.target.checked, props)} />);

const ListForm: React.FunctionComponent<IFieldForm> = (props) => (
  <EuiSelect
    {...props.item.form.params}
    value={getValue(props)}
    onChange={(e) => onChange(e.target.value, props)} />);

const IntervalForm: React.FunctionComponent<IFieldForm> = (props) => {
  const [interval, setInterval] = useState<string>(getValue(props));
  const [invalid, setInvalid] = useState(false);
  useEffect(() => {
    if (validate(interval)) {
      setInvalid(false);
      getValue(props) !== interval && onChange(interval, props);
    } else {
      setInvalid(true);
      deleteChange(props);
    }
  }, [interval])
  return (
    <>
      <EuiFieldText
        fullWidth
        value={interval}
        isInvalid={invalid}
        onChange={e => setInterval(e.target.value)}
      />
      {invalid && <EuiTextColor color='danger'>Invalid cron schedule expressions</EuiTextColor>}
    </>
  );
}

const ArrayForm: React.FunctionComponent<IFieldForm> = (props) => {
  const [list, setList] = useState(JSON.stringify(getValue(props)));

  useEffect(() => {
    checkErrors();
  }, [list]);

  const checkErrors = () => {
    try {
      const parsed = JSON.parse(list);
      onChange(parsed, props);
    } catch (error) {
      const options: UIErrorLog = {
        context: `${FieldForm.name}.checkErrors`,
        level: UI_LOGGER_LEVELS.ERROR as UILogLevel,
        severity: UI_ERROR_SEVERITIES.UI as UIErrorSeverity,
        error: {
          error: error,
          message: error.message || error,
          title: error.message || error,
        },
      };

      getErrorOrchestrator().handleError(options);
    }
  }
  return (
    <EuiCodeEditor
      mode='javascript'
      height='50px'
      width='100%'
      value={list}
      onChange={setList}
   />
  );
}

//#endregion

//#region Helpers

const getValue = ({ item, updatedConfig }: IFieldForm) => typeof updatedConfig[item.setting] !== 'undefined'
  ? updatedConfig[item.setting]
  : item.value;

const onChange = (value: string | number | boolean | [], props: IFieldForm) => {
  const { updatedConfig, setUpdatedConfig, item } = props;
  if(!_.isEqual(item.value,value)){
    setUpdatedConfig({
      ...updatedConfig,
      [item.setting]: value,
    })
  }else{
    deleteChange(props);
  }
}

const deleteChange = (props: IFieldForm) => {
  const { updatedConfig, setUpdatedConfig, item } = props;
  const newConfig = { ...updatedConfig };
  delete newConfig[item.setting];
  setUpdatedConfig(newConfig);
}

//#endregion
