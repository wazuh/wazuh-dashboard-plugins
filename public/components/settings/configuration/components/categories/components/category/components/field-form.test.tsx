/*
 * Wazuh app - React test for FieldForm component.
 *
 * Copyright (C) 2015-2022 Wazuh, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Find more information about this on the LICENSE file.
 *
 */

import React from 'react';
import { FieldForm } from './field-form';
import { ISetting } from '../../../../../configuration';
import { mount } from 'enzyme';

describe('FieldForm component', () => {
  it('renders correctly to match the snapshot', () => {
    const item: ISetting = {
      setting: 'string',
      value: 'boolean',
      description: 'string',
      category: 'string',
      name: 'string',
      form: { type: 'text', params: {} }
    };
    const updatedConfig = {};
    const setUpdatedConfig = jest.fn();

    const wrapper = mount(
      <FieldForm item={item} updatedConfig={updatedConfig} setUpdatedConfig={setUpdatedConfig} />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
