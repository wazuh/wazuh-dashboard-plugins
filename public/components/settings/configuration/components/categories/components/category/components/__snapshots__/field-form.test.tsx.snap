// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldForm component renders correctly to match the snapshot 1`] = `
<FieldForm
  item={
    Object {
      "category": "string",
      "description": "string",
      "form": Object {
        "params": Object {},
        "type": "text",
      },
      "name": "string",
      "setting": "string",
      "value": "boolean",
    }
  }
  setUpdatedConfig={[MockFunction]}
  updatedConfig={Object {}}
>
  <TextForm
    item={
      Object {
        "category": "string",
        "description": "string",
        "form": Object {
          "params": Object {},
          "type": "text",
        },
        "name": "string",
        "setting": "string",
        "value": "boolean",
      }
    }
    setUpdatedConfig={[MockFunction]}
    updatedConfig={Object {}}
  >
    <EuiFieldText
      fullWidth={true}
      onChange={[Function]}
      value="boolean"
    >
      <EuiFormControlLayout
        fullWidth={true}
      >
        <div
          className="euiFormControlLayout euiFormControlLayout--fullWidth"
        >
          <div
            className="euiFormControlLayout__childrenWrapper"
          >
            <EuiValidatableControl>
              <input
                className="euiFieldText euiFieldText--fullWidth"
                onChange={[Function]}
                type="text"
                value="boolean"
              />
            </EuiValidatableControl>
            <EuiFormControlLayoutIcons />
          </div>
        </div>
      </EuiFormControlLayout>
    </EuiFieldText>
  </TextForm>
</FieldForm>
`;
